### PR TITLE
New and improved gulp4 workflow

### PR DESCRIPTION
**Pull request by 0dp**
All dev assets now live in ./src
That means there is no ./sass folder anymore. Instead it lives in ./src/scss/

There is currently no need for theme.css/js and theme.min.css/js
Instead files are being built according to task. So `npm run build` will create minified and uglyfied versions of theme.css/js

Understrap now support ES6
Components can be created in ./src/js/components and should be imported in ./src/js/theme.js
```
// Here you can import custom reusable components
import './components/component.js';
```
In ./src/js/theme.js you can also include regular js and jQuery scripts
```
// Here you can import custom scripts
import './custom-scripts.js';
```

You do not need to make workarounds to use $ instead of jQuery. This should be handled automatically by webpack. Yes there is webpack and gulp 😬 

### Available Commands

> Please make sure you are running the latest version of gulp-cli (2.0.1)
> Please verify with `gulp -v`

`npm run start` watch, compile and reload. replaces `gulp watch bs`
`npm run build` same as `gulp dist`
`gulp styles` compile scss to css with sourcemap
`gulp styles --prod` compile scss to css minify etc. without sourcemap
`gulp images` compress imagess and move to ./images
`gulp scripts` compile js, es6 components and minify
`gulp clean` delete dist
`gulp dist` build dist
`gulp watchForChanges` replaces gulp watch since watch is reserved
`gulp copyAssets` replaces copy-assets

### CODE REVIEW

**Comment by 0dp** [comment]
^ I don't know why gulpfile.js is a conflict. gulpfile.js does not exist any more. Instead it is called. gulpfile.babel.js

@holger1411

**Comment by Thomas-A-Reinert** [comment]
@holger1411  - I´d really recommend to get this one merged asap!

Not only that I´d personally really love to use (and test-run) it for my current project, but also to respect @0dp ´s work and dedication. He did it once some time ago, now did it the second time and I guess he won´t do a third time...

Would be a real improvement also.

**Comment by 0dp** [comment]
Considering the first one was merged, into upstream but later taken out because of lag of testing. The old gulp4 setup is now stale, so something like this is needed.

I foresee lots of big reports coming in soon if we don't take migration to gulp4 serious; Or webpack.

**Comment by axlright** [comment]
@0dp Please let me know what I can do to help test the gulp4 branch. I've checked it out on my local development server.

Coupla questions:

1) You mention:
>All dev assets now live in ./src That means there is no ./sass folder anymore. Instead it lives in ./src/scss/
But I still see a `./sass` directory. Want to make sure I'm looking at things right.

2) You mention `npm run build` but I'm not able to execute that command - I get `npm ERR! missing script: build`

3) You mention:
> I don't know why gulpfile.js is a conflict. gulpfile.js does not exist any more. Instead it is called. gulpfile.babel.js
But I don't see a `gulpfile.babel.js` on the gulp4 branch at https://github.com/understrap/understrap/tree/gulp4

4) When I run `npm install` it runs `gulp copy-assets` though you mention `gulp copyAssets` replaces `copy-assets` so again just want to make sure I'm doing things right.

lease let me know specifically what I can do beyond this to help test. Maybe I'm supposed to check out the specific #883 PR but I don't know if I know how to do that...

Thanks again for all your work and Happy New Year and let me know what I can do.

**Comment by axlright** [comment]
@0dp Okay nevermind I'm an idiot. I need to check out and test https://github.com/0dp/understrap/tree/master not https://github.com/understrap/understrap/tree/gulp4

Hang on sorry for the stupid.

**Comment by axlright** [comment]
@0dp Okay got https://github.com/0dp/understrap/tree/master running locally. I can't fork it because I already have the mothership Understrap forked.

In your `package.json` on line 7 you have `"postinstall": "gulp copAssets",` and changing that typo to `"postinstall": "gulp copyAssets",` got `npm install` to run with 0 errors or vulnerabilities. Nice!

I'll continue testing. Let me know anything specific to hit. I'll go through your setup notes above.

**Comment by axlright** [comment]
@0dp I ran `npm run build` and then copied all of the files in the `./dist` directory into the parent `understrap` theme directory of a local WordPress testing install and then ran `npm install` and got the following error:

```
> gulp copyAssets

[14:08:26] Requiring external module @babel/register
/Users/alexwright/Sites/understrap-demo/web/wp-content/themes/understrap/gulpfile.babel.js:2
import { src, dest, watch, series, parallel } from 'gulp';
^^^^^^

SyntaxError: Unexpected token import
    at new Script (vm.js:51:7)
    at createScript (vm.js:136:10)
    at Object.runInThisContext (vm.js:197:10)
    at Module._compile (internal/modules/cjs/loader.js:618:28)
    at Module._compile (/Users/alexwright/Sites/understrap-demo/web/wp-content/themes/understrap/node_modules/pirates/lib/index.js:83:24)
    at Module._extensions..js (internal/modules/cjs/loader.js:665:10)
    at Object.newLoader [as .js] (/Users/alexwright/Sites/understrap-demo/web/wp-content/themes/understrap/node_modules/pirates/lib/index.js:88:7)
    at Module.load (internal/modules/cjs/loader.js:566:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:506:12)
    at Function.Module._load (internal/modules/cjs/loader.js:498:3)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! understrap@0.8.9 postinstall: `gulp copyAssets`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the understrap@0.8.9 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```
However, directly copying all the files except `.git .github .gitignore` and `./node-modules` into the same WordPress testing install then running `npm install` works with no errors, so there may be something with `npm run build` that doesn't include necessary files to run as a development theme.

**Comment by 0dp** [comment]
It looks like the errors you are getting are related to Babel missing and the es6 syntax.

Can you confirm you have the .babelrc file ?

Also try and delete the package-lock.js file and node modules and try and run `npm install` again

Also for good measurements what is the output of `gulp -v`

**Comment by axlright** [comment]
Heya Johan Happy New Year and glad to hear there's a hangover involved 💪 ‼️ 

I do have the `.babelrc` that contains

```
{
  "presets": ["@babel/preset-env"]
}
```
Deleting `package.json` and `node-modules` and re-running `npm install` didn't do the trick. Which is sad.

Checking the version of gulp outputs maybe a clue (or maybe not) [as discussed here](https://github.com/gulpjs/gulp/issues/1631)

```
newbian-three:understrap alexwright$ gulp --version
[20:54:06] Failed to load external module babel-register
[20:54:06] Failed to load external module babel-core/register
[20:54:06] Failed to load external module babel/register
[20:54:06] CLI version 1.4.0
[20:54:06] Local version 4.0.0
```

**Comment by axlright** [comment]
Thanks @0dp and I hope your hangover is better.

You are 💯 on the gulp-cli version 2. Everything is now working perfectly.

What can I test for you? I'll try to get my uncss integration working from #858 and otherwise put your new gulp4 workflow into production on a live site. If there's anything else I can test, please holla.

-Alex

**Comment by 0dp** [comment]
Hahaha yeah we back to normal 😆

Awesome, good to hear!
I don't think there is much else to test right now.
We just need green light from @holger1411 to merge.

**Comment by axlright** [comment]
@0dp I added your gulp4 workflow to a child theme. Coupla questions:

- Are there plans to add a non-minified version of the css file as well as css map to the original .scss file? Both are features I end up using a lot.

- I couldn't get the Sass to compile due to an error that's somewhat unrelated but didn't come up before with the previous Sass compiler. So I'm wondering what is different about the compiler. The error has to do with `"env(safe-area-inset-right)" is not a number for 'max'` in this css value:
```
.fancybox-caption {
    padding: 75px max(44px, env(safe-area-inset-right)) max(25px, env(safe-area-inset-bottom)) max(44px, env(safe-area-inset-left));
}
```
This issue is documented at [sass/libsass/issues/2521](https://github.com/sass/libsass/issues/2521) and [sass/sass#2378](https://github.com/sass/sass/issues/2378) but the previous Sass compiler for Understrap didn't take issue.

- You say that "In ./src/js/theme.js you can also include regular js and jQuery scripts" but I'm curious about having an entire directory where multiple scripts can get added to the final `theme.js`. We used a directory in the old gulpfile.js like so 

```
/ Run: 
// gulp scripts. 
// Uglifies and concat all JS files into one
gulp.task( 'scripts', function() {
	var scripts = [

		// Start - All BS4 stuff
		paths.dev + '/js/bootstrap4/bootstrap.js',
		// End - All BS4 stuff

		paths.dev + '/js/skip-link-focus-fix.js',

		// Start - Owl Carousel script, comment out if not using
		// paths.dev + 'js/owl.carousel.js',
		// Start - Owl Carousel scripts

		// Start - Custom scripts
		paths.custom_js + '/include/*.js'
		// End - Custom scripts

	];
```
So adding scripts to `paths.custom_js + '/include/*.js'` was helpful because we could drop in re-used scripts and / or edit individual scripts that would then add get batch added to the final js file. Maybe this isn't a good way to do Javascript. I'm not that smart.

Thanks!

**Comment by 0dp** [comment]
@axlright sup, did that answer your questions or?
@holger1411 @Thomas-A-Reinert had a chance to test?

**Comment by axlright** [comment]
@0dp I have not been able to implement, but I will and can continue to tinker around. Please let me know if there's anything else that can be tested.

**Comment by 0dp** [comment]
@axlright if my answers to your questions was sufficient, and you have tested those then no. Everything should be in finest order. I really appreciate your testing but I would, as last time, like to see more engagement from the rest of the team in testing and developing this new set of tools. 🙏

**Comment by Thomas-A-Reinert** [comment]
Hey @0dp I fear i´m not a great help atm. 
Customer demanded to fork his theme on the stable branch. And I fear I´m still not git´ish enough to testmerge both. Sorry. Would love to get so far but..

My current job (never thought I´d do that EVER, especially since it´s Javascript/Typescript based) as the lead voice-application Google-Actions Dev for Germanys leading radiostation WDR (something like the BBC for the UK) ist killing me. Experiencing new skills and expertise and getting paid for it and really, really enjoying it.
But that´s a hard time contributing to OS projects which I´d love to.

So please don´t count me in in the near future. Trying to contribute/commit whatever I can, but again: Do Not Count On Me. I just lack the time to run some test-environment.

**Comment by 0dp** [comment]
Sounds amazing though, congrats!

**Comment by Mte90** [comment]
there are conflicts with this pr

**Comment by Mte90** [comment]
I will start to develop using that branch to test it :-)

**Comment by 0dp** [comment]
> I will start to develop using that branch to test it :-)

I would really appreciate feedback

**Comment by Mte90** [comment]
I got:
```
[17:50:54] Task never defined: copAssets
[17:50:54] To list available tasks, try running: gulp --tasks
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! understrap@0.8.9 postinstall: `gulp copAssets`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the understrap@0.8.9 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above```

File: package.json [hunk 1]
```diff
@@ -1,13 +1,12 @@
...
-  "version": "0.8.3",
+  "version": "0.8.9",
   "description": "WordPress Theme framework",
   "main": "index.js",
   "scripts": {
-    "postinstall": "gulp copy-assets"
-  },
-  "engines": {
-    "npm": ">=2.1.8"
+    "postinstall": "gulp copAssets",
```
Comment (Mte90) [inline package.json]: There is a typo here, it is `copyAssets`

Comment (0dp) [reply package.json]: After correcting the postinstall command to gulp copyAssets, no other error?

**Comment by Mte90** [comment]
I started doing a theme with this fork https://github.com/Mte90/daniele.tech-2019

**Comment by 0dp** [comment]
That's awesome 🙏

**Comment by axlright** [comment]
Is there any word from @holger1411 on when this new Gulp4 workflow can get integrated into the production version of the repo?

**Comment by 0dp** [comment]
No news ☹️

**Comment by holger1411** [comment]
Sorry for my late reply guys!
Ive got  alot of merge conflicts and it seems i´am not very familiar with gulp 4 (i try to fidle it out but received alot of follow up errors.) Seems the gulp 4 syntax is quite different from the gulp 3 syntax.

Maybe you could help me out with it?

**Comment by 0dp** [comment]
Yes of course. So let's start off with the merge conflicts and errors you're facing.

**Comment by Mte90** [comment]
Just a suggestion to bump the bootstrap version to 4.3.1 because the previos versions are vulnerable to a XSS and github report and alert in the repo.

**Comment by 0dp** [comment]
Word! I'll look into it asap

**Comment by 0dp** [comment]
My fork is updated now.

@holger1411 You don't have to merge into master but could we at least try and update the gulp4 branch with my fork please?

**Comment by Mte90** [comment]
@0dp you forgot to change also the issue of package.json that I reported 2 weeks ago :-)

**Comment by 0dp** [comment]
Argh, damn it!

I'll fix it, in a minute. Just frying some mushrooms

**Comment by 0dp** [comment]
@Mte90 I just had a look, I can't seem to find any traces of copAssets typo. Are you downloading the fork from my page or through here?

Please cop it from here, https://github.com/0dp/understrap

File: package.json (hunk 1 above)
Comment (Mte90) [reply package.json]: seems that now is fixed

**Comment by Mte90** [comment]
nevermind the latest commit fixed it :-)

**Comment by Mte90** [comment]
I discovered that this pr use https://www.npmjs.com/package/font-awesome that is at 4.7.0 version updated 2 years ago but the last font awesome version is 5.7.2 version.

**Comment by Mte90** [comment]
this are the changes to use the latest fontawesome with npm https://github.com/Mte90/daniele.tech-2019/commit/e30fb11d9a2c9a516024126f88af831f5dd4c61a

**Comment by 0dp** [comment]
Correct! There is a long running discussion on this

**Comment by danemorgan** [comment]
@0dp I believe we are already on a Gulp 4 workflow? Is this an improvement to the one we are currently on or was this PR superseded by another that was accepted in its place?

**Comment by 0dp** [comment]
It looks like this PR is completely screwed up. I can't make heads or tails of it. Probably because my fork has been reset. So I'll close this.



> @0dp I believe we are already on a Gulp 4 workflow? Is this an improvement to the one we are currently on or was this PR superseded by another that was accepted in its place?

We do. This was the second attempt to include gulp4 into upstream back then. It had a slightly different approach to the current approach, which is an adaption/patch of the gulp3 workflow to work with gulp4.

If i remember correctly, it tried to utilize the gulp4 features like series and parallel tasks to improve speed. It used flags for building production code instead of separate tasks. So you didn't have different minified and unminified files. It introduced the first steps for uncss and relied more on postcss because node-sass.

It had revision history for cache busting, and a webpack config so you could load plugins and filters beyond that of gulp (because sometimes gulp plugins are just shallow wrappers). But it still tried to be backward compatible with the old build system.

Lastly, i think it cleaned up the directory structure, so all dev stuff was inside the src folder.

 Ahh yes, the memories of a forgotten time.

**Pull request closed without merging**